File: Packages/RecordingKit/Sources/RecordingKit/GIFTiming.swift
import Foundation
import CoreGraphics

public enum GIFTiming {
    /// Sample timestamps (seconds) for converting a clip to GIF at `fps`.
    /// Always at least one frame.
    public static func frameTimes(duration: Double, fps: Int) -> [Double] {
        guard duration > 0, fps > 0 else { return [0] }
        let step = 1.0 / Double(fps)
        let count = max(1, Int(duration * Double(fps)))
        return (0..<count).map { Double($0) * step }
    }

    /// Aspect-preserving downscale to `maxWidth`; never upscales.
    public static func outputSize(source: CGSize, maxWidth: CGFloat) -> CGSize {
        guard source.width > maxWidth, source.width > 0 else { return source }
        let scale = maxWidth / source.width
        return CGSize(width: maxWidth, height: (source.height * scale).rounded())
    }
}

File: windows/src/BetterScreenshot.Recording/GIFTiming.cs
using BetterScreenshot.Core;

namespace BetterScreenshot.Recording;

/// <summary>Pure frame-time and downscale math for MP4→GIF export.</summary>
public static class GIFTiming
{
    /// <summary>Frame sample times [0, 1/fps, 2/fps, …) covering <paramref name="duration"/> seconds.</summary>
    public static IReadOnlyList<double> FrameTimes(double duration, int fps)
    {
        var times = new List<double>();
        if (fps <= 0 || duration <= 0) return times;
        double step = 1.0 / fps;
        int count = (int)Math.Floor(duration * fps + 1e-9);
        for (int i = 0; i < count; i++) times.Add(i * step);
        return times;
    }

    /// <summary>Aspect-preserving downscale so width ≤ <paramref name="maxWidth"/>; never upscales.</summary>
    public static PxSize OutputSize(PxSize source, double maxWidth)
    {
        if (source.Width <= maxWidth) return source;
        double scale = maxWidth / source.Width;
        return new PxSize(maxWidth, Math.Round(source.Height * scale));
    }
}
